### ISSUE TITLE
Class not colored when it is static in C#

### ISSUE THREAD

**Issue by cadilhac** [tmont/sunlight#4]
new MyClass().MyMethod();
-> MyClass is correctly colored

MyClass.Initialize();
-> MyClass is not colored

Visual Studio, for example, will color the class in both cases.

So could a named ident rule help in getting MyClass colored? For example, a rule that says that if a word begins with an upper letter and ends with a dot or "<" (for generics)... ?

**Comment by tmont** [comment]
Unfortunately there isn't a reliable way to tell when an expression is a call to a static method. For example, the following is ambiguous:

``` csharp
class Console {
  public static void WriteLine() {}
}
class Foo {
  public Thing Console { get; set; }
  public void Write() {
    Console.WriteLine();
  }
}
```

Without performing static analysis, it's impossible to tell whether `Console.WriteLine()` is a reference to the static class `Console` or the local property `Console`. In these cases, I chose to leave them unhighlighted. Basically static analysis is outside the scope of this library.

In some cases (Scala's case classes, for one), I've performed rudimentary static analysis by just keeping track of what classes have been declared previously and using that list as a last resort for highlighting situations like this. For C#, though, it would be more complicated, due to having to keep track of properties, fields, methods, etc. And it's imperfect anyway, since the order that things are declared in would be significant.

In short, it can't be done reliably.

**Comment by cadilhac** [comment]
I agree with you, but is there a way to define a custom named indent rule just for this case? I know that it will work with the code I display and I understand it won't with other pieces of code. In other terms how to set this rule so that it takes into account the case of the first letter and the last char being . or <

**Comment by tmont** [comment]
I whipped this up just now. It'll handle the following:

``` csharp
Foo.Bar(); //"Foo" will be highlighted
Foo<Bar>.Baz(); //"Foo" will be highlighted
Foo.Bar; //"Foo" will NOT be highlighted
```

It doesn't handle classes with more than one generic type parameter, e.g. `Foo<Bar, Baz>.Qux();`. It also won't handle type parameters on the method name, e.g. `Foo.Bar<Baz>()`. But it could be generalized to do so.

Paste the following as another array item in the `namedIdentRules.custom` section of `sunlight.csharp.js`"

``` javascript
//static class names in static method calls, e.g. "Foo" in Foo.Bar()
function(context) {
    var ident = context.tokens[context.index].value;
    if (!/^[A-Z]/.test(ident)) {
        //must start with a capital letter
        return false;
    }

    //handles "Foo" in Foo.Bar()
    var nonGeneric = [
        sunlight.util.whitespace,
        { token: "operator", values: ["."] },
        sunlight.util.whitespace,
        { token: "ident" },
        sunlight.util.whitespace,
        { token: "punctuation", values: ["("] }
    ];

    //handles "Foo" in Foo<Baz>.Bar()
    var generic = [
        sunlight.util.whitespace,
        { token: "operator", values: ["<"] },
        sunlight.util.whitespace,
        { token: "ident" },
        sunlight.util.whitespace,
        { token: "operator", values: [">"] },
        sunlight.util.whitespace,
        { token: "operator", values: ["."] },
        sunlight.util.whitespace,
        { token: "ident" },
        sunlight.util.whitespace,
        { token: "punctuation", values: ["("] }
    ];

    return sunlight.util.createProceduralRule(context.index + 1, 1, nonGeneric)(context.tokens) ||
        sunlight.util.createProceduralRule(context.index + 1, 1, generic)(context.tokens);
},
```

**Comment by tmont** [comment]
Here's a commit illustrating where the code goes: https://github.com/tmont/sunlight/commit/9ec44c32146d3c996a0490b2a376ddca8cbc4825

**Comment by cadilhac** [comment]
Great, thank you. I will have a look.

However, I thought these kind of rules were made to be added outside of your javascript files. There is no way to do that at config time, from my js file?

**Comment by cadilhac** [comment]
I tried your solution. While good, it makes strange that Foo.Bar does not have Foo highlighted just after a line with Foo.Bar() where Foo is highlighted. The effect is bad and at final, it is maybe better when they were not colored at first... On my side I will allow Foo.Bar.

But I still don't know how to add this function without touching your script. Passing it when calling highlightAll would be great.

**Comment by tmont** [comment]
"Foo" in `Foo.Bar` is not highlighted because that snippet is written to only highlight static method calls, not property expressions. You can change it to whatever you need it to do. Or tell me what your use case is and I can try make it work for you.

The language configurations are global to the library (i.e. every time you highlight something), so you can't just pass in a local configuration parameter to change them. You could register a new language to replace an old one, though.

**Comment by patricklafrance** [comment]
I would like to achieve the same thing. I tried your snippet, and it works great if I  have something like Foo.Bar() but if I have Foo.SomethingElse.Bar() then Foo is not highlighted. In my case, the rule should be something like that:

If it starts with a capital letter, it is the first word of an expression and it is followed by a dot, it should be a named-ident. For exemple:

If I have Foo.SomethingElse.Bar(), only Foo should be a named-ident.
If I have var  myVar = Foo.SomethingElse.Bar() only Foo should be a named-ident.

Do you have any ideas about how do implement something like that ?

**Comment by tmont** [comment]
Yeah, that should be possible. I'll try and whip something up that accomplishes that tomorrow.

**Comment by tmont** [comment]
@patricklafrance Take a look at ca015bb (in `sunlight.csharp.js`). I didn't spend a lot of time verifying that it doesn't screw something else up, and it almost certainly is going to miss some corner cases, but I think it accomplishes what you were looking for.

Here's a screenshot of what it'll highlight:

![sunlight-csharp-static](https://cloud.githubusercontent.com/assets/188562/2716378/e9986c98-c528-11e3-8809-05c22a5a08f7.png)

Hope that helps.

**Comment by patricklafrance** [comment]
Thank you for very much for your help. This is working great.

Is it possible to make it work for generics ? Here's a screenshot

![statics_with_generics](https://cloud.githubusercontent.com/assets/794579/2724133/688801a2-c59d-11e3-94ba-de3cbd44ac78.PNG)

**Comment by tmont** [comment]
@patricklafrance This should do the trick: https://github.com/tmont/sunlight/commit/9c9310c5f496602add05c6aef96f2ccfff18b882

Screenshot:

![sunlight-charp-static2](https://cloud.githubusercontent.com/assets/188562/2737458/e28b3e18-c67b-11e3-93f3-b3703ddbd334.png)

**Comment by tmont** [comment]
Oops, there was a bug in that commit. Use the file at cabc03a: [sunlight.charp.js](https://github.com/tmont/sunlight/blob/cabc03abef4712f0fd255e7aeb48c2d55f99c09c/src/lang/sunlight.csharp.js)

**Comment by patricklafrance** [comment]
Thats works perfectly!!!

I added the following the handle nested generics:

(token.name === "operator" && token.value === "<<") ||
(token.name === "operator" && token.value === ">>") ||

Thank you very much for your help, your syntax highlighter is awesome, I never thought I could find a plugin that achieve that quality of highlighting!

**Comment by tmont** [comment]
:smiley:
